CI/wininstaller/installer.iss

```diff
@@ -0,0 +1,733 @@
+  // Check if the installer is running on a 64-bit system
+  if DirExists(ExpandConstant('{win}\syswow64')) then
+  begin
+    if ExpandConstant('{#InstallerArch}') = 'x64' then
+      // For 64-bit installer, return the 64-bit Program Files directory
+      Result := ExpandConstant('{commonpf64}')
+    else
+      // For 32-bit installer on 64-bit system, return the 32-bit Program Files directory
+      Result := ExpandConstant('{commonpf32}');
+  end
+  else
+    // On 32-bit systems, always return the 32-bit Program Files directory
+    Result := ExpandConstant('{commonpf32}');
+end;
+
+
+function GetDefaultDir(Default: String): String;
+begin
+  if IsAdmin then
+    // Default to Program Files for admins
+    Result := GetCommonProgramFilesDir + '\VCMI'
+  else
+    // Default to User AppData for non-admin users
+    Result := GlobalUserAppdataFolder + '\VCMI';
+end;
+
+
+function GetUserDocsFolder: String;
+var
+  UserDocsPath: String;
+  OriginalUserName: String;
+  CurrentSessionUserName: String;
+begin
+  // Retrieve the current username from the session
+  CurrentSessionUserName := '\' + GlobalUserName + '\';
+
+  // Retrieve the original username
+  OriginalUserName := '\' + GetUserNameString + '\';
+
+  // Expand the {userdocs} constant
+  UserDocsPath := ExpandConstant('{userdocs}');
+
+  // Replace the original username with the current session username in the user docs path
+  StringChangeEx(UserDocsPath, OriginalUserName, CurrentSessionUserName, True);
+  
+  // Return the modified user docs folder path
+  Result := UserDocsPath;
+end;
+
+
+function GetUserAppdataFolder: String;
+var
+  UserAppdataPath: String;
+  OriginalUserName: String;
+  CurrentSessionUserName: String;
+begin
+  // Retrieve the current username from the session
+  CurrentSessionUserName := '\' + GlobalUserName + '\';
+
+  // Retrieve the original username
+  OriginalUserName := '\' + GetUserNameString + '\';
+
+  // Expand the {userappdata} constant
+  UserAppdataPath := ExpandConstant('{userappdata}');
+
+  // Replace the original username with the current session username in the user Appdata path
+  StringChangeEx(UserAppdataPath, OriginalUserName, CurrentSessionUserName, True);
+  
+  // Return the modified user docs folder path
+  Result := UserAppdataPath;
+end;
+
+
+function GetUserDesktopFolder(Default: String): String;
+var
+  UserDesktopPath: String;
+  OriginalUserName: String;
+  CurrentSessionUserName: String;
+begin
+  // Retrieve the current username from the session
+  CurrentSessionUserName := '\' + GlobalUserName + '\';
+
+  // Retrieve the original username
+  OriginalUserName := '\' + GetUserNameString + '\';
+
+  // Expand the {userdesktop} constant
+  UserDesktopPath := ExpandConstant('{userdesktop}');
+
+  // Replace the original username with the current session username
+  StringChangeEx(UserDesktopPath, OriginalUserName, CurrentSessionUserName, True);
+
+  // Return the modified user desktop folder path
+  Result := UserDesktopPath;
```

Comment:
this code is used 3 times and differs only by the constant used :)

Reply:
Will change it to this:

```
function GetUserFolderPath(Constant: String): String;
var
  FolderPath: String;
  OriginalUserName: String;
  CurrentSessionUserName: String;
begin
  // Retrieve the current username from the session
  CurrentSessionUserName := '\' + GlobalUserName + '\';

  // Retrieve the original username
  OriginalUserName := '\' + GetUserNameString + '\';

  // Expand the specified constant
  FolderPath := ExpandConstant(Constant);

  // Replace the original username with the current session username in the path
  StringChangeEx(FolderPath, OriginalUserName, CurrentSessionUserName, True);

  // Return the modified folder path
  Result := FolderPath;
end;

// Specific functions for user folders
function GetUserDocsFolder: String;
begin
  Result := GetUserFolderPath('{userdocs}');
end;

function GetUserAppdataFolder: String;
begin
  Result := GetUserFolderPath('{userappdata}');
end;

function GetUserDesktopFolder(Default: String): String;
begin
  Result := GetUserFolderPath('{userdesktop}');
end;
```